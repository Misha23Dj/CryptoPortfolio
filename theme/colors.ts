export type Theme = {
  bg: string;
  text: string;
  muted: string;
  border: string;
  positive: string;
  negative: string;
  card: string;
  kpiLabel: string;
  kpiValue: string;
  barStyle: "light" | "dark";
  statusBarBg: string;
};

export const light: Theme = {
  bg: "#ffffff",
  text: "#0f172a",
  muted: "#64748b",
  border: "#e5e7eb",
  positive: "#16a34a",
  negative: "#dc2626",
  card: "#f8fafc",
  kpiLabel: "#64748b",
  kpiValue: "#0f172a",
  barStyle: "dark",
  statusBarBg: "#ffffff",
};

export const dark: Theme = {
  bg: "#0b1220",
  text: "#e5e7eb",
  muted: "#9aa4b2",
  border: "#1f2937",
  positive: "#22c55e",
  negative: "#f87171",
  card: "#111827",
  kpiLabel: "#9aa4b2",
  kpiValue: "#e5e7eb",
  barStyle: "light",
  statusBarBg: "#0b1220",
};
