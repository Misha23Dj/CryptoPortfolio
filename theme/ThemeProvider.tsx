import React, { createContext, useContext, useMemo, useState } from "react";
import { useColorScheme } from "react-native";
import { dark, light, type Theme } from "./colors";

type ThemeMode = "light" | "dark" | "system";

const ThemeCtx = createContext<Theme>(light);
const ThemeModeContext = createContext<{
  mode: ThemeMode;
  setMode: (m: ThemeMode) => void;
}>({
  mode: "system",
  setMode: () => {},
});

export const useTheme = () => useContext(ThemeCtx);
export const useThemeMode = () => useContext(ThemeModeContext);

type Props = { children: React.ReactNode };

const ThemeProvider: React.FC<Props> = ({ children }) => {
  const [mode, setMode] = useState<ThemeMode>("system");
  const scheme = useColorScheme();

  const effective = mode === "system" ? scheme ?? "light" : mode;
  const theme = useMemo(
    () => (effective === "dark" ? dark : light),
    [effective]
  );

  return (
    <ThemeModeContext.Provider value={{ mode, setMode }}>
      <ThemeCtx.Provider value={theme}>{children}</ThemeCtx.Provider>
    </ThemeModeContext.Provider>
  );
};

export default ThemeProvider;
