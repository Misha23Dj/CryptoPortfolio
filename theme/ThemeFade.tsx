import React, { useEffect, useRef } from "react";
import { StyleSheet } from "react-native";
import Animated, {
  Easing,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from "react-native-reanimated";
import { useTheme, useThemeMode } from "../theme/ThemeProvider";

const ThemeFade: React.FC = () => {
  const theme = useTheme();
  const { mode } = useThemeMode();
  const opacity = useSharedValue(0);
  const prev = useRef(mode);

  useEffect(() => {
    if (prev.current !== mode) {
      opacity.value = 1;
      opacity.value = withTiming(0, {
        duration: 250,
        easing: Easing.out(Easing.cubic),
      });
      prev.current = mode;
    }
  }, [mode]);

  const animatedStyle = useAnimatedStyle(() => ({ opacity: opacity.value }));

  return (
    <Animated.View
      pointerEvents="none"
      style={[
        StyleSheet.absoluteFillObject,
        { backgroundColor: theme.bg, zIndex: 10 },
        animatedStyle,
      ]}
    />
  );
};

export default ThemeFade;
