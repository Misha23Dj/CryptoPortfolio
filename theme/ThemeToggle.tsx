import React from "react";
import { Pressable, StyleSheet, Text, View } from "react-native";
import { useTheme, useThemeMode } from "./ThemeProvider";

const MODES = ["light", "dark", "system"] as const;
type Mode = (typeof MODES)[number];

const ThemeToggle: React.FC = () => {
  const theme = useTheme();
  const { mode, setMode } = useThemeMode();

  return (
    <View
      style={[
        styles.wrap,
        { borderColor: theme.border, backgroundColor: theme.card },
      ]}
    >
      {MODES.map((m) => {
        const selected = mode === m;
        return (
          <Pressable
            key={m}
            onPress={() => setMode(m as Mode)}
            style={[
              styles.chip,
              {
                backgroundColor: selected ? theme.text : "transparent",
                borderColor: theme.border,
              },
            ]}
          >
            <Text
              style={{
                color: selected ? theme.bg : theme.text,
                fontWeight: "600",
              }}
            >
              {m[0].toUpperCase() + m.slice(1)}
            </Text>
          </Pressable>
        );
      })}
    </View>
  );
};

const styles = StyleSheet.create({
  wrap: {
    flexDirection: "row",
    gap: 8,
    padding: 6,
    borderWidth: StyleSheet.hairlineWidth,
    borderRadius: 12,
    alignSelf: "flex-end",
    marginRight: 16,
    marginBottom: 8,
  },
  chip: {
    paddingVertical: 6,
    paddingHorizontal: 10,
    borderWidth: StyleSheet.hairlineWidth,
    borderRadius: 10,
  },
});

export default ThemeToggle;
