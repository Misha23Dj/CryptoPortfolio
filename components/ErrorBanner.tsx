import { useTheme } from "@/theme/ThemeProvider";
import React from "react";
import { StyleSheet, Text, View } from "react-native";

const ErrorBanner: React.FC<{ message: string }> = ({ message }) => {
  const theme = useTheme();
  return (
    <View
      style={[
        styles.wrap,
        { backgroundColor: theme.negative + "11", borderColor: theme.negative },
      ]}
    >
      <Text style={[styles.text, { color: theme.negative }]}>{message}</Text>
    </View>
  );
};

const styles = StyleSheet.create({
  wrap: {
    marginHorizontal: 16,
    padding: 10,
    borderRadius: 10,
    borderWidth: StyleSheet.hairlineWidth,
    marginBottom: 12,
  },
  text: { fontWeight: "600" },
});

export default ErrorBanner;
