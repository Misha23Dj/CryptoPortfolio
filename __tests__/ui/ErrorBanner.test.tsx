import ErrorBanner from "@/components/ErrorBanner";
import ThemeProvider from "@/theme/ThemeProvider";
import { light as lightTheme } from "@/theme/colors";
import { render } from "@testing-library/react-native";
import React from "react";

const renderWithTheme = (ui: React.ReactElement) =>
  render(<ThemeProvider>{ui}</ThemeProvider>);

it("renders the message", () => {
  const { getByText } = renderWithTheme(<ErrorBanner message="Boom" />);
  expect(getByText("Boom")).toBeTruthy();
});

it("applies themed colors", () => {
  const { getByText, getByTestId } = renderWithTheme(
    <ErrorBanner message="Boom" />
  );
  expect(getByText("Boom")).toHaveStyle({ color: lightTheme.negative });
  expect(getByTestId("error-banner")).toHaveStyle({
    backgroundColor: `${lightTheme.negative}11`,
    borderColor: lightTheme.negative,
  });
});

it("matches snapshot", () => {
  const { toJSON } = renderWithTheme(
    <ErrorBanner message="Something went wrong" />
  );
  expect(toJSON()).toMatchSnapshot();
});
