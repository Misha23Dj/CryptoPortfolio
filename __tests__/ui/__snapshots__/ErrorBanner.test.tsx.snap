// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<View
  style={
    [
      {
        "borderRadius": 10,
        "borderWidth": 0.5,
        "marginBottom": 12,
        "marginHorizontal": 16,
        "padding": 10,
      },
      {
        "backgroundColor": "#dc262611",
        "borderColor": "#dc2626",
      },
    ]
  }
  testID="error-banner"
>
  <Text
    style={
      [
        {
          "fontWeight": "600",
        },
        {
          "color": "#dc2626",
        },
      ]
    }
  >
    Something went wrong
  </Text>
</View>
`;
